papermill/execute.py

```diff
@@ -13,81 +11,69 @@
-    os.remove(tmp_path)
+    nb = load_notebook_node(notebook)
 
+    # Parameterize the Notebook.
+    if parameters:
+        _parameterize_notebook(nb, parameters)
 
-def _parameterize_notebook(notebook_path, output_path, parameters):
+    # Execute the Notebook.
+    t0 = time.time()
+    processor = ExecutePreprocessor(
+        timeout=None,
+        kernel_name=kernel_name or nb.metadata.kernelspec.name
+    )
+    processor.preprocess(nb, {})
+    duration = time.time() - t0
 
-    if not parameters:
-        # No parameters, just copy the notebook.
-        shutil.copyfile(notebook_path, output_path)
-        return
+    # Record specified environment variable values.
+    nb.metadata.papermill['parameters'] = parameters
+    nb.metadata.papermill['environment_variables'] = _fetch_environment_variables()
+    nb.metadata.papermill['metrics']['duration'] = duration
+
+    # Write Notebook to disk.
+    write_ipynb(nb, output)
 
-    # Pull out variable names and values from the parameters argument.
-    param_content = "# Parameters\n"
-    for var, val in parameters.items():
-        if isinstance(val, string_types):
-            val = '"%s"' % val  # TODO: Handle correctly escaping input strings.
-        param_content += '%s = %s\n' % (var, val)
 
-    # Find `parameters` cell and replaced with our parameters.
-    nb = load_notebook_node(notebook_path)
+def _parameterize_notebook(nb, parameters):
+    param_content = _build_parameter_code(nb, parameters)
+    param_cell_index = _find_parameters_index(nb)
 
-    parameters_index = find_parameters_index(nb)
-    old_parameters = nb.cells[parameters_index]
-    before = nb.cells[:parameters_index]
-    after = nb.cells[parameters_index + 1:]
+    # Remove the old cell and replace it with a new one containing parameter content.
+    old_parameters = nb.cells[param_cell_index]
+    before = nb.cells[:param_cell_index]
+    after = nb.cells[param_cell_index + 1:]
     newcell = nbformat.v4.new_code_cell(source=param_content)
     newcell.metadata['tags'] = old_parameters.metadata.tags
     nb.cells = before + [newcell] + after
 
-    # Apply papermill metadata
-    nb.metadata.papermill['parameters'] = parameters
-    write_ipynb(nb, output_path)
-
-
-def _execute_notebook(notebook_path, output_path):
 
-    t0 = time.time()
-    exit_code = subprocess.call(
-        ['jupyter', 'nbconvert', '--to', 'notebook',  '--execute',
-         '--NbConvertApp.output_base={}'.format(notebook_path),
-         '--ExecutePreprocessor.timeout=-1',
-         '--ClearOutputPreprocessor.enabled=True', notebook_path]
+def _build_parameter_code(nb, parameters):
+    kernel_name = nb.metadata.kernelspec.name
+    kernel_lang = nb.metadata.kernelspec.language
+    if kernel_name in _parameter_code_builders:
+        return _parameter_code_builders[kernel_name](parameters)
+    elif kernel_lang in _parameter_code_builders:
+        return _parameter_code_builders[kernel_lang](parameters)
```

Comment:
Great to see you looked into the kernelspec to find the language for the parameter insertion.